### ISSUE TITLE
write_table does not respect indices dtype

### ISSUE THREAD

**Issue by hkad98** [apache/arrow#36589]
### Describe the bug, including details regarding any error messages, version, and platform.

When I am working with a table that has a field of dictionary type, I try to store it as a parquet using [write_table](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.write_table.html) and then load it using [read_table](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.read_table.html) I noticed that write_table does not respect dictionary indices type. Please take a look at the reproducer below.

PyArrow: 12.0.1
Python: 3.11.1
Platform: MacOS, 13.4.1 (22F82)

Reproducer:
```python
import pyarrow as pa
import pyarrow.parquet as pq

table = pa.Table.from_arrays([["a", "b", None, "d"]], schema=pa.schema([pa.field("random",pa.dictionary(pa.int8(), pa.string()))]))

"""
pyarrow.Table
random: dictionary<values=string, indices=int8, ordered=0>
----
random: [  -- dictionary:
["a","b","d"]  -- indices:
[0,1,null,2]]
"""

pq.write_table(table, "example.parquet")
table_cpy = pq.read_table("example.parquet")

"""
pyarrow.Table
random: dictionary<values=string, indices=int32, ordered=0>
----
random: [  -- dictionary:
["a","b","d"]  -- indices:
[0,1,null,2]]
"""
```

### Component(s)

Parquet, Python

**Comment by AlenkaF** [comment]
I am not sure but looking at the Parquet docs _I think_ this is the design of the Parquet file format (https://github.com/apache/parquet-format#types).

What you could do is to use the schema when reading the parquet file in `read_table` also.

**Comment by hkad98** [comment]
IMHO, int8 should not be the issue. If I use the int8 type, it is stored and loaded properly.

```python 
import pyarrow as pa
import pyarrow.parquet as pq

table = pa.Table.from_arrays([[0,1,2,3]], schema=pa.schema([pa.field("example", pa.int8())]))

"""
pyarrow.Table
example: int8
----
example: [[0,1,2,3]]
"""

pq.write_table(table, "example.parquet")
table_cpy = pq.read_table("example.parquet")

"""
pyarrow.Table
example: int8
----
example: [[0,1,2,3]]
"""
```

**Comment by AlenkaF** [comment]
Oh, of course - you are correct.

I see the same behaviour with legacy reader in pyarrow.parquet or in the datasets:

```python
```

**Comment by AlenkaF** [comment]
Oh, of course - you are correct.

I see the same behaviour with legacy reader in pyarrow.parquet or in the datasets API:

```python
>>> my_schema = pa.schema([
...     pa.field("random", pa.dictionary(pa.int8(), pa.string())),
...     pa.field("int8", pa.int8()),
... ])
>>> table = pa.table([["a", "b", None, "d"], [1, 2, 3, 4]], schema=my_schema)
>>> table
pyarrow.Table
random: dictionary<values=string, indices=int8, ordered=0>
int8: int8
----
random: [  -- dictionary:
["a","b","d"]  -- indices:
[0,1,null,2]]
int8: [[1,2,3,4]]
>>> ds.write_dataset(table, "example", format="parquet")
>>> dataset = ds.dataset("example", format="parquet")
>>> dataset.to_table()
pyarrow.Table
random: dictionary<values=string, indices=int32, ordered=0>
int8: int8
----
random: [  -- dictionary:
["a","b","d"]  -- indices:
[0,1,null,2]]
int8: [[1,2,3,4]]
```

```python
>>> pq.write_table(table, "example.parquet")
pq.read_table("example.parquet", use_legacy_dataset=True)>>> pq.read_table("example.parquet", use_legacy_dataset=True)
<stdin>:1: FutureWarning: Passing 'use_legacy_dataset=True' to get the legacy behaviour is deprecated as of pyarrow 8.0.0, and the legacy implementation will be removed in a future version.
pyarrow.Table
random: dictionary<values=string, indices=int32, ordered=0>
int8: int8
----
random: [  -- dictionary:
["a","b","d"]  -- indices:
[0,1,null,2]]
int8: [[1,2,3,4]]
```

I have also found an open issue about this behaviour with a good answer: https://github.com/apache/arrow/issues/30302

**Comment by AlenkaF** [comment]
Duplicate of https://github.com/apache/arrow/issues/30302

**Comment by idavi-bcs** [comment]
Unfortunately it looks like this is a well-known bug at this point, reported multiple times (#36589, #30302, #27616).  However, I want to point out an important impact that I haven't seen mentioned yet.  I have a large data table of int8 categoricals (genotype data), but it still fits comfortably in memory, and can easily be written to Parquet.  But I cannot *read* the Parquet file back into memory, because now it takes 5 times as much space (int32 + int8 in memory simultaneously as Pandas tries to cast back to int8).  So my data is effectively lost.

In other words, this is not just a performance bug -- it can actually cause data loss in the case of large tables!
